### PR TITLE
Fix Subscribe panic and add a test which would have caught it

### PR DESCRIPTION
**Pull request by nshalman**
## Description

This PR creates a new test for the Subscribe endpoint as well as fixing a current bug which causes panics when that RPC is used.

## Why is this needed

Fixes: #69

## How Has This Been Tested?

```console
~/Workspace/hegel % go test -run TestSubscribe -count 2000 ./grpc-server/...
ok  	github.com/tinkerbell/hegel/grpc-server	15.102s

~/Workspace/hegel % git revert --no-edit 88728c4
Auto-merging grpc-server/grpc_server.go
[nshalman/subscribe-fix-and-test 1d53e08] Revert "subLock needs to be initialized"
 Date: Fri Jan 7 13:14:24 2022 -0500
 1 file changed, 1 deletion(-)

~/Workspace/hegel % go test -run TestSubscribe ./grpc-server/...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x107525e]

goroutine 104 [running]:
sync.(*RWMutex).Lock(0xc00031d080)
<snip>
```

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Comment by nshalman** [comment]
Thanks again to @mmlb for most of this test harness, and for the pointers for the further cleanup of the tests in https://github.com/tinkerbell/hegel/compare/669989512f74834d3a41cbcc271cf03cb152c630..f0ae94cd34199f0d1d1ce3c90e64dfb8ad9111a5

**Comment by mmlb** [comment]
> Thanks again to @mmlb for most of this test harness, and for the pointers for the further cleanup of the tests in [`669989512f74834d3a41cbcc271cf03cb152c630..f0ae94cd34199f0d1d1ce3c90e64dfb8ad9111a5` (compare)](https://github.com/tinkerbell/hegel/compare/669989512f74834d3a41cbcc271cf03cb152c630..f0ae94cd34199f0d1d1ce3c90e64dfb8ad9111a5)

np. now that I look again the `WaitGroup` is probably unnecessary and I forgot to drop it

**Comment by nshalman** [comment]
> now that I look again the `WaitGroup` is probably unnecessary and I forgot to drop it

Good catch. I will clean that up.

**Review by tobert — APPROVED** [review]
I read it over and think I understand what's happening and it looks fine to me. Thanks for adding this!

File: grpc-server/grpc_helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// Logs the given message and marks the test as failed.
+func (zt *zapT) Errorf(format string, args ...interface{}) {
+	zt.mu.RLock()
+	if !zt.done {
+		zt.T.Logf(format, args...)
+	}
+	zt.mu.RUnlock()
+}
+
+// serveAndDial starts 2 grpc services.
```
Comment (mmlb) [inline grpc-server/grpc_helpers_test.go:48]: should be `startServersAndConnectClient`

File: grpc-server/subscribe_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		assert.NoError(t, err)
+		assert.NotNil(t, w)
+
+		hw, err := w.Recv()
+		assertGRPCError(t, err, terr)
+		assert.Nil(t, hw)
+	})
+
+	t.Run("good", func(t *testing.T) {
+		// id := uuid.Must(uuid.NewRandom()).String()
```
Comment (mmlb) [inline grpc-server/subscribe_test.go:46]: should be deleted

File: grpc-server/subscribe_test.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		id := "bufconn"
+		data := map[string]string{}
+		value := fmt.Sprintf(`{"id": "%s", "ip": "%s"}`, id, id)
+		payload := fmt.Sprintf(`%s=%s`, id, value)
+		data[id] = value
```
Comment (mmlb) [inline grpc-server/subscribe_test.go:51]: could be:
```suggestion
		value := fmt.Sprintf(`{"id": "%s", "ip": "%s"}`, id, id)
		payload := fmt.Sprintf(`%s=%s`, id, value)
		data := map[string]string{
			id: value,
		}
```
but not a big deal to change really

**Review by mmlb — CHANGES_REQUESTED** [review]
just a few minor nits the only one I'd really like to see fixed is the comment on `startServersAndConnectClient`

File: grpc-server/grpc_helpers_test.go (hunk 1 above)
Comment (nshalman) [reply grpc-server/grpc_helpers_test.go:48]: And the rest of the comment updated too. 😄

File: grpc-server/grpc_server.go [hunk 4]
```diff
@@ -61,9 +61,11 @@ func NewServer(l log.Logger, hc hardware.Client) (*Server, error) {
 		}
 	}
 
+	var m sync.RWMutex
 	s := &Server{
 		log:            l,
 		hardwareClient: hc,
+		subLock:        &m,
```
[the remark this answers is not in the document]
Comment (nshalman) [reply grpc-server/grpc_server.go]: Done

File: grpc-server/grpc_helpers_test.go (hunk 1 above)
Comment (nshalman) [reply grpc-server/grpc_helpers_test.go]: Done

File: grpc-server/subscribe_test.go (hunk 2 above)
Comment (nshalman) [reply grpc-server/subscribe_test.go]: Done

File: grpc-server/subscribe_test.go (hunk 3 above)
Comment (nshalman) [reply grpc-server/subscribe_test.go]: Done

File: grpc-server/subscribe_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+		}()
+
+		hw := cacher.Hardware{}
+		assert.NoError(t, w.RecvMsg(&hw))
+		orig := hw.JSON
+		close(initialPushes)
+
+		// ok now that w.RecvMsg has processed at least one push we can run the actual test.
+		// We need to accept plain old $id as the value because the initialPush go routine
+		// could have snuck in a couple extra pushes while we get back from the w.RecvMsg call
```
Comment (nshalman) [inline grpc-server/subscribe_test.go:86]: This comment should be updated too

**Review by mmlb — APPROVED** [review]
:shipit:

**Pull request merged**